### PR TITLE
Move Zen discovery into module

### PR DESCRIPTION
**Pull request by ywelsch**
This PR moves zen discovery to a module, stripping core from the zen discovery functionality.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/gateway/GatewayService.java [hunk 1]
```diff
@@ -202,7 +202,7 @@ private void performStateRecovery(boolean enforceRecoverAfterTime, String reason
                 threadPool.schedule(recoverAfterTime, ThreadPool.Names.GENERIC, () -> {
                     if (recovered.compareAndSet(false, true)) {
                         logger.info("recover_after_time [{}] elapsed. performing state recovery...", recoverAfterTime);
-                        gateway.performStateRecovery(recoveryListener);
+                        gateway.performStateRecovery(recoveryListener, Math.max(1, recoverAfterMasterNodes));
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/gateway/GatewayService.java]: higher up in this file we refer to min master nodes explicitly with a vague todo which you missed because it's not done via a code link. I think it's OK to leave the hard coded string for now (as we plan to move this things to another place anyway), but I think it's good to change the comment to explain that you can't use the settings object because it's a module and that there are plans to fix it.

File: modules/discovery-zen/src/test/java/org/elasticsearch/discovery/zen/ClusterDisruptionTests.java [hunk 2]
```diff
@@ -73,7 +71,7 @@
  */
 @ClusterScope(scope = Scope.TEST, numDataNodes = 0, transportClientRatio = 0, autoMinMasterNodes = false)
 @TestLogging("_root:DEBUG,org.elasticsearch.cluster.service:TRACE")
-public class ClusterDisruptionIT extends AbstractDisruptionTestCase {
+public class ClusterDisruptionTests extends AbstractDisruptionTestCase {
```
Comment (bleskes) [inline modules/discovery-zen/src/test/java/org/elasticsearch/discovery/zen/ClusterDisruptionTests.java]: why did you rename the the IT extension? it's still an integration test?

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedDiscoveryPlugin.java [hunk 3]
```diff
@@ -104,8 +103,8 @@ public void close() throws IOException {
     public Settings additionalSettings() {
         // For 5.0, the hosts provider was "zen", but this was before the discovery.zen.hosts_provider
         // setting existed. This check looks for the legacy zen, and sets the file hosts provider if not set
-        String discoveryType = DiscoveryModule.DISCOVERY_TYPE_SETTING.get(settings);
-        if (discoveryType.equals("zen")) {
+        if (DiscoveryModule.DISCOVERY_TYPE_SETTING.exists(settings) == false ||
```
Comment (bleskes) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedDiscoveryPlugin.java]: why the extra check for exists? isn't the get enough?

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java [hunk 4]
```diff
@@ -593,7 +594,8 @@ private NodeAndClient buildNode(int nodeId, long seed, Settings settings,
             .put("node.name", name)
             .put(NodeEnvironment.NODE_ID_SEED_SETTING.getKey(), seed);
 
-        final boolean usingSingleNodeDiscovery = DiscoveryModule.DISCOVERY_TYPE_SETTING.get(finalSettings.build()).equals("single-node");
+        final boolean usingSingleNodeDiscovery = DISCOVERY_TYPE_SETTING.exists(settings) &&
```
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: why the extra exists?

**Review by bleskes — APPROVED** [review]
LGTM.

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java (hunk 4 above)
Comment (ywelsch) [inline test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: The issue here is that we are not accessing the true "final" settings of the node. These are just the settings explicitly passed to the node. If the Zen-Disco plugin is installed, it will extend these settings by setting the `discovery.type` to `test-zen`. The default discovery implementation is "single-node" now, which means that if the settings we are using here do not have an explicit ``discovery.type` setting, then the code here thinks that we will be using `single-node`, which is wrong.
It's a bit of a chicken-egg problem. We can only know about the actual setting of "discovery.type" after we've constructed the mock-node, but here we also want to influence the settings of the mock node before constructing it. In theory we could construct a temporary MockNode but that is also very heavy-weight.

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedDiscoveryPlugin.java (hunk 3 above)
Comment (bleskes) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedDiscoveryPlugin.java]: I got confused and I see why you added this. I think it merits an extra comment indicating that at this point the discovery type setting is not defaulted to zen. Also, I think we can remove this clause completely in 6.0. This deprecation happened in 5.1.1 (#21036)

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java (hunk 4 above)
Comment (bleskes) [inline test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: I see. Super confusing :) can you add a comment that single usage has to be preconfigured? alternatively we can check for ZenDiscovery in the plugin list and the fact that discovery type is not configured/configured to zen. Maybe that's more intuitive as it is a positive test of "is zen discovery active". wdyt?

File: plugins/discovery-file/build.gradle [hunk 5]
```diff
@@ -53,6 +54,7 @@ integTestCluster {
     dependsOn setupSeedNodeAndUnicastHostsFile
     clusterName = 'discovery-file-test-cluster'
     extraConfigFile 'discovery-file/unicast_hosts.txt', srcUnicastHostsFile
+    module project(':modules:discovery-zen')
```
Comment (rjernst) [inline plugins/discovery-file/build.gradle]: This means discovery-file is going to load it's entire own copy of zen. I don't think that is what we want...

File: modules/discovery-zen/build.gradle [hunk 6]
```diff
@@ -17,19 +17,17 @@
...
 
-import java.util.List;
+compileJava.options.compilerArgs << "-Xlint:-cast,-deprecation,-rawtypes,-try,-unchecked"
+compileTestJava.options.compilerArgs << "-Xlint:-cast,-deprecation,-rawtypes,-try,-unchecked"
 
-/**
- * A pluggable provider of the list of unicast hosts to use for unicast discovery.
- */
-public interface UnicastHostsProvider {
+// somehow both packaged version and main source set end up on classpath, tripping jarhell check
```
Comment (rjernst) [inline modules/discovery-zen/build.gradle]: Jarhell occurs because this module was added directly as a dependency for test framework. To be truly independent, we need the test framework to not depend on this.

**Review by rjernst — COMMENTED** [review]
I'm not sure about this as it is currently. Having a test framework dep on this new module mean that it is not really modularized (because everything pulls in test framework).

File: plugins/discovery-file/build.gradle (hunk 5 above)
Comment (ywelsch) [inline plugins/discovery-file/build.gradle]: No, that's not true. This just ensures that the zen discovery module is installed as well on the deployed node for the REST test. The REST tests for plugins use the stripped down distribution `integ-test-zip`, not including the modules. The discovery-file module cannot be meaningfully tested without zen-discovery though.

**Comment by ywelsch** [comment]
> Having a test framework dep on this new module mean that it is not really modularized (because everything pulls in test framework).

I disagree with this view. Modules are bundled with the Elasticsearch distribution, i.e., are part of the core bundle (in contrast to plugins). The `test-framework` provides means to test the core bundle, which also includes the `discovery-zen` module. The reason to include `discovery-zen` into `test-framework` is to be able to run Fantasyland tests with a proper discovery implementation. We agree already that Fantasyland tests are not equivalent to the real thing, and for the real thing we have proper isolation, so I don't see any harm done here.

**Comment by nik9000** [comment]
If you decide that the test framework should depend on the module then please make sure this all works in Eclipse. Eclipse doesn't have "test" and "production" configurations for projects. It flattens all the dependencies together. It is one of the "fun" "quirks" that we Eclipse users have worked around for so long we barely think about anymore. Mostly it doesn't matter. But in some cases like this you can get dependency cycles in Eclipse that aren't cycles in other Gradle or IntelliJ. I believe this might be one of those cases.

**Comment by rjernst** [comment]
@ywelsch Is this PR something you would still like to get in?

**Comment by ywelsch** [comment]
As it's not clear yet how this fits into the upcoming Zen disco changes and a future upgrade path, I'm closing this for now (but might revisit this later).
